Add node class

vhaller opened the pull request:
All recent changes merged to one pull request.

vhaller committed:
Added the function node:new() and a corresponding unit test.

vhaller committed:
added the function node:add() and a corresponding unit test.

vhaller committed:
added the function node:last_child() and a corresponding unit test.

vhaller committed:
added the function node:level() and a matching unit test.

vhaller committed:
All recent changes in one commit

kulla commented on src/Node.lua:
```diff
@@ -0,0 +1,22 @@
+function Node:add_child(child)
+  self.children[#self.children + 1] = child
+  child.parent = self
+  child.index  = #self.children
```
<del>Why do you set the `index` attribute to the children's index? Is this attribute used somewhere?!</del>


Delete this line. It's really ugly code :smile:

vhaller replied:
Right! I didn't notice.

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,39 @@
+local root, node_1, node_2, node_3 = Node:new(), Node:new(), Node:new(), Node:new()
+root:add_child(node_1)
+root:add_child(node_2)
+node_2:add_child(node_3)
```
Use a `setUp()` function here (http://luaunit.readthedocs.io/en/latest/#grouping-tests-setup-teardown-functionality )

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,39 @@
+node_2:add_child(node_3)
+
+function TestNode:test_new()
+  luaunit.assertTrue(getmetatable(root) == Node)
```
Use [assertIs](http://luaunit.readthedocs.io/en/latest/#assertIs):

```
luaunit.assertIs(getmetatable(root), Node)
```

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,39 @@
+function TestNode:test_new()
+  luaunit.assertTrue(getmetatable(root) == Node)
+  luaunit.assertNotNil(root)
+  luaunit.assertNil(root.parent)
```
Missing:

```
luaunit.assertItemsEquals(root.children, {})
```

See http://luaunit.readthedocs.io/en/latest/#assertItemsEquals

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,39 @@
+  luaunit.assertTrue(node_1.parent == root)
+  luaunit.assertTrue(node_2.parent == root)
+  luaunit.assertTrue(node_3.parent == node_2)
+  luaunit.assertTrue(#root.children == 2 and #node_2.children == 1)
```
Use `assertEquals` and write it in two lines...

vhaller committed:
Added the function Node:find(predicate_fn) to Node.lua and a corresponding unit test to Node_test.lua.

kulla commented:
@vhaller: Have a look on [how to format a git commit message](http://chris.beams.io/posts/git-commit/) :smile:

kulla commented on test/Node_test.lua:
```diff
@@ -36,4 +36,15 @@ function TestNode:test_level()
 end
 
+function TestNode:test_find()
+  luaunit.assertTrue(root:find(function(node) return node == node_3 end) == node_3)
```
Use [`assertIs`](http://luaunit.readthedocs.io/en/latest/#assertIs) here... (and in the following lines)

vhaller committed:
Change formal inaccuracies.

Use assertIs()/assertEquals()/assertItemsEquals() instead of assertTrue().
Remove the unused entry `index` from the body of Node:add_child().

vhaller committed:
Add a new test for Node:add_child().

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+  luaunit.assertIs(root:last_child(), node_2)
+  luaunit.assertIs(node_2:last_child(), node_3)
+
+  luaunit.assertFalse(node_1:last_child())
```
Use `assertIsNil` here...

vhaller committed:
Change assertFalse() to assertNil()

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+
+  luaunit.assertNotNil(root)
+
+  luaunit.assertNil(root.parent)
```
<del>`luaunit.assertItemsEquals(root.children, {})` is missing</del>

vhaller replied:
When I add this assertion, the test fails:

## Failed tests:

1) Test_Node.test_new
.\test\Node_test.lua:18: Contents of the tables are not identical:
Expected:
{}
Actual:
<table: 00630460> {
    <table: 00630488> {
        children=<table: 006304B0> {},
        parent=<table: 00630438> {children=<table: 00630460>}
    },
    <table: 006304D8> {
        children=<table: 00630500> {<table: 00630528> {children=<table: 00630550> {}, parent=<table: 006304D8>}},
        parent=<table: 00630438>
    }
}
stack traceback:
        .\test\Node_test.lua:18: in function 'Test_Node.test_new'

kulla replied:
<del>Stimmt. Add `local root = Node:new()` at the beginning of the test...</del>

vhaller replied:
Alright, thanks.

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+root:add_child(node_1)
+root:add_child(node_2)
+node_2:add_child(node_3)
+
```
<del>Use a `setUp()` function here (http://luaunit.readthedocs.io/en/latest/#grouping-tests-setup-teardown-functionality )</del>

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+
+local TestNode = {}
+
+local root, node_1, node_2, node_3 = Node:new(), Node:new(), Node:new(), Node:new()
```
<del>use variable names `node1`, `node2`, `node3`</del>

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+  luaunit.assertIs(node_1.parent, root)
+  luaunit.assertIs(node_2.parent, root)
+  luaunit.assertIs(node_3.parent, node_2)
+  luaunit.assertItemsEquals(root.children, {node_1, node_2})
```
<del>Add

```
luaunit.assertItemsEquals(node_1.children, {})
luaunit.assertItemsEquals(node_2.children, { node_3 })
luaunit.assertItemsEquals(node_3.children, {})</del>
```

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+function TestNode:test_add()
+  luaunit.assertIs(node_1.parent, root)
+  luaunit.assertIs(node_2.parent, root)
+  luaunit.assertIs(node_3.parent, node_2)
```
eine Leerzeile hier (ich weiß, dass ist ne ziemliche Kleinigkeit)... :smile:

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+  luaunit.assertItemsEquals(root.children, {node_1, node_2})
+
+  luaunit.assertEquals(#root.children, 2)
+  luaunit.assertEquals(#node_2.children, 1)
```
Diese zwei Zeilen können gelöscht werden, da sie bereits durch

```
luaunit.assertItemsEquals(root.children, {node_1, node_2})
```

abgedeckt werden...

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+  luaunit.assertNil(root.parent)
+end
+
+function TestNode:test_add()
```
<del>Add `luaunit.assertIsNil(root.parent)`</del>

vhaller replied:
Isn't it there already?

kulla replied:
`test_add` hasn't this test already... (each test case needs all assertations in order have a full coverage even if there are overlaps between tests)

vhaller replied:
I understand.

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+  luaunit.assertIs(root:find(function(node) return node == node_3 end), node_3)
+  luaunit.assertIs(root:find(function(node) return node.parent ~= nil end), node_1)
+
+  luaunit.assertNotIs(root:find(function(node) return node:level() == 1 end), node_2)
```
<del>Mach hier ein `node:level() == 2` raus (muss dann `node_2` zurückgeben)</del>

vhaller replied:
In Zeile 48?

kulla replied:
Passt

kulla reviewed (changes requested):
Last things which are needed before the merge...

kulla commented:
At the end only one function is missing:

- [ ] `Node:find_parent()`

vhaller commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+  luaunit.assertNil(root.parent)
+end
+
+function TestNode:test_add()
```
To TestNode:test_new() or to TestNode:test_add()?

kulla commented on test/Node_test.lua:
```diff
@@ -0,0 +1,56 @@
+  luaunit.assertIs(root:find(function(node) return node == node_3 end), node_3)
+  luaunit.assertIs(root:find(function(node) return node.parent ~= nil end), node_1)
+
+  luaunit.assertNotIs(root:find(function(node) return node:level() == 1 end), node_2)
```
ah, ne, passt hier...

kulla replied:
Passt

kulla reviewed (commented):
AW

vhaller committed:
Implement minor corrections

Summary:
- Change the names of variables to the format nodex from node_x, where x is a number in {1,2,3}.
- Add the function setUp(), which provides each unit test with the necessary variables; each of the test functions may adress these variables by using the prefix self.
- Add an assertion which tests if root initially has no children.
- Remove two unnecessary lines from and add three assertions to TestNode:test_add()

vhaller committed:
Add function find_parent and a corresponding unit test
